### PR TITLE
[Hardware][CPU] Support AWQ for CPU backend

### PR DESCRIPTION
**Pull request by bigPYJ1151**
This PR added a new ```ipex_quant``` configuration to support various weight-only quantization method with IPEX for the CPU backend.

AWQ is supported for now, a basic test is included.

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by robertgshaw2-neuralmagic** [comment]
Thanks!

**Comment by bigPYJ1151** [comment]
Hi @mgoin , would you please help to review this PR? Because I added a platform filter in ```awq_marlin``` compatibility checking. Thanks! :)

File: vllm/model_executor/layers/quantization/ipex_quant.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+            WeightOnlyQuantizedLinear.from_weight(
+                layer.qweight,
+                layer.scales,
+                layer.qzeros,
+                layer.qweight.size(0),
+                layer.ipex_output_size,
+                qconfig=qconfig,
+                bias=bias,
+                group_size=self.quant_config.group_size,
+                quant_method=1
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/ipex_quant.py:128]: Is there a more informative value that could be used here?

File: vllm/model_executor/layers/quantization/ipex_quant.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
+        super().process_weights_after_loading(layer=layer)
+
+        bias = layer.bias if not layer.skip_bias_add else None
+
+        import intel_extension_for_pytorch as ipex
+
+        weight_dtype = ipex.quantization.WoqWeightDtype.INT4
+        lowp_mode = ipex.quantization.WoqLowpMode.INT8
+        act_quant_mode = ipex.quantization.WoqActQuantMode.PER_BATCH
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/ipex_quant.py:107]: Can you describe these settings? It seems like this may be quantizing activations as well?

File: vllm/model_executor/layers/quantization/ipex_quant.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        return [
+            "quant_config.json",
+            "quantize_config.json",
+        ]
+
+    @classmethod
+    def from_config(cls, config: Dict[str, Any]) -> "IPEXConfig":
+        method = cls.get_from_keys(config, ["quant_method"]).lower()
+        weight_bits = cls.get_from_keys(config, ["w_bit", "bits"])
+        group_size = cls.get_from_keys(config, ["q_group_size", "group_size"])
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/ipex_quant.py:62]: Can you check that the number of bits and group size is supported by this implementation?

File: vllm/model_executor/layers/quantization/ipex_quant.py (hunk 1 above)
Comment (bigPYJ1151) [reply vllm/model_executor/layers/quantization/ipex_quant.py]: This argument is defined as int for now. I added a dict mapping to make it more informative. (AWQ=1, GPTQ=2)

File: vllm/model_executor/layers/quantization/ipex_quant.py (hunk 2 above)
Comment (bigPYJ1151) [reply vllm/model_executor/layers/quantization/ipex_quant.py]: Yes, the ```lowp_mode``` means the ```compute type```, using int8 for FMA instructions with better performance. The activation is dynamic per-token quantized to int8 with the setting ```PER_BATCH```.

File: vllm/model_executor/layers/quantization/ipex_quant.py (hunk 3 above)
Comment (bigPYJ1151) [reply vllm/model_executor/layers/quantization/ipex_quant.py:74]: Added a checking for ```weight_bits=4```. There is no restriction for group_size.

**Comment by bigPYJ1151** [comment]
Hi @mgoin , do you have any further comments about this? I think this PR is ready to go :)

BTW, seems the CPU CI on the main branch is broken due to a new MOE model test and the lack of the AZP epilogue. I disabled the compressed-tensor test and will fix it recently.

File: vllm/model_executor/layers/quantization/ipex_quant.py (hunk 3 above)
Comment (mgoin) [reply vllm/model_executor/layers/quantization/ipex_quant.py:74]: Thanks

File: vllm/model_executor/layers/quantization/ipex_quant.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+    def __init__(self, quant_config: IPEXConfig):
+        self.quant_config = quant_config  # type: ignore
+
+    def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
+        super().process_weights_after_loading(layer=layer)
+
+        bias = layer.bias if not layer.skip_bias_add else None
+
+        import intel_extension_for_pytorch as ipex
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/ipex_quant.py:115]: Could you put a try-catch around the ipex import to provide a better instructive error message to the user? It is also important to enforce versioning if needed. 
Something like this deepspeed backend https://github.com/vllm-project/vllm/blob/151ef4efd2fb52554f4d30408aca619e181ea751/vllm/model_executor/layers/quantization/deepspeedfp.py#L144-L153

**Review by mgoin — COMMENTED** [review]
This looks to be in a good place now, thanks! I just have a nit on UX @bigPYJ1151 
Please let me know if I can assist with the compressed-tensors test as we will want to enable that again.

**Review by mgoin — APPROVED** [review]
LGTM! Thanks for the updates, let's see if it can get green

**Comment by bigPYJ1151** [comment]
Hi @mgoin , looks the tests get green, please help to merge this PR, thanks :)

I'll open a PR for the AZP soon, will let you know for review.
